api/cluster/resource/templater.go

```diff
@@ -918,13 +921,38 @@ func (t *InferenceServiceTemplater) getResourceLimitsAndEnvVars(
 			resourceRequest.MemoryRequest, t.deploymentConfig.UserContainerMemoryLimitRequestFactor,
 		)
 	}
-	return limits, envVars, nil
+
+	// Add user-configured secrets as env vars to the list of env vars to be created
+	mergedEnvVars = MergeEnvVars(mergedEnvVars, secrets.ToKubernetesEnvVars(secretName))
+
+	return limits, mergedEnvVars, nil
+}
+
+// MergeEnvVars merges multiple sets of environment variables and return the merging result.
+// All the EnvVars passed as arguments will be not mutated.
+// EnvVars to the right have higher precedence.
+func MergeEnvVars(left []corev1.EnvVar, rightEnvVars ...[]corev1.EnvVar) []corev1.EnvVar {
+	for _, right := range rightEnvVars {
+		envIndexMap := make(map[string]int, len(left)+len(right))
+		for index, ev := range left {
+			envIndexMap[ev.Name] = index
+		}
+		for _, add := range right {
+			if index, exist := envIndexMap[add.Name]; exist {
+				left[index].Value = add.Value
+				left[index].ValueFrom = add.ValueFrom
+			} else {
+				left = append(left, add)
+			}
+		}
+	}
+	return left
 }
```

Comment:
This is basically just a copy of whatever already has been implemented [here](https://github.com/caraml-dev/merlin/blob/main/api/models/env_var.go#L75). Basically I've refactored the code a little to perform merging of lists of env vars with the final `corev1.EnvVar` datatype (used by k8s, kserve, etc.) instead of the intermediate `model.EnvVar` datatype (only used by Merlin) in order to simplify the merging process. Some env vars do not come from the `model.EnvVar` datatype (like env vars with references to secrets) and it's weird to perform merging only with the `model.EnvVar` datatype. 

This simplification also removes the need for the `ParseEnvVars` helper function below.